Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
#

In [12]:
data = pd.read_csv("labeledTrainData.tsv", sep="	")

In [14]:
data.head()

id  sentiment                                             review
0  5814_8          1  With all this stuff going down at the moment w...
1  2381_9          1  \The Classic War of the Worlds\" by Timothy Hi...
2  7759_3          0  The film starts with a manager (Nicholas Bell)...
3  3630_4          0  It must be assumed that those who praised this...
4  9495_8          1  Superbly trashy and wondrously unpretentious 8...

In [15]:
data.describe()

sentiment
count  3842.000000
mean      0.500781
std       0.500064
min       0.000000
25%       0.000000
50%       1.000000
75%       1.000000
max       1.000000

In [16]:
del data['id']

In [17]:
data.head()

sentiment                                             review
0          1  With all this stuff going down at the moment w...
1          1  \The Classic War of the Worlds\" by Timothy Hi...
2          0  The film starts with a manager (Nicholas Bell)...
3          0  It must be assumed that those who praised this...
4          1  Superbly trashy and wondrously unpretentious 8...

In [18]:
data['sentiment'].value_counts()

sentiment
1    1924
0    1918
Name: count, dtype: int64

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3842 entries, 0 to 3841
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  3842 non-null   int64 
 1   review     3842 non-null   object
dtypes: int64(1), object(1)
memory usage: 60.2+ KB


In [20]:
data.describe( )

sentiment
count  3842.000000
mean      0.500781
std       0.500064
min       0.000000
25%       0.000000
50%       1.000000
75%       1.000000
max       1.000000

In [21]:
data['review'][1]

'\\The Classic War of the Worlds\\" by Timothy Hines is a very entertaining film that obviously goes to great effort and lengths to faithfully recreate H. G. Wells\' classic book. Mr. Hines succeeds in doing so. I, and those who watched his film with me, appreciated the fact that it was not the standard, predictable Hollywood fare that comes out every year, e.g. the Spielberg version with Tom Cruise that had only the slightest resemblance to the book. Obviously, everyone looks for different things in a movie. Those who envision themselves as amateur \\"critics\\" look only to criticize everything they can. Others rate a movie on more important bases,like being entertained, which is why most people never agree with the \\"critics\\". We enjoyed the effort Mr. Hines put into being faithful to H.G. Wells\' classic novel, and we found it to be very entertaining. This made it easy to overlook what the \\"critics\\" perceive to be its shortcomings."'

In [22]:
data['review'][0]

"With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.<br /><br />Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.<br /><br />The actual feature film bit when it finally sta

DATA PREPROCESSING

In [23]:
# As the text contains puncuation, quotes that can affect the accuracy these has to be removed

TEXT Cleaning

In [24]:
import re
def clean_text(text):
    text = re.compile('<.*#?/>').sub('', text)
    text = re.sub(r'[^\w\s]','',text, re.UNICODE)
    text = text.lower()
    text = text.strip()
    return text

In [25]:
data['review'] = data['review'].apply(clean_text)

In [26]:
# To Remove Special characters

In [27]:
def remove_char(text):
    x=''
    for i in text:
        if i.isalnum():
            x=x+i
        else:
            x=x+ " "
    return x

In [28]:
data['review'] = data['review'].apply(remove_char)

In [29]:
# Converting the text to lower case

In [30]:
def convert_lower(text):
    return text.lower()

In [31]:
data['review'] = data['review'].apply(convert_lower)

In [32]:
data.head()

sentiment                                             review
0          1  with all this stuff going down at the moment w...
1          1  the classic war of the worlds by timothy hines...
2          0  the film starts with a manager nicholas bell g...
3          0  it must be assumed that those who praised this...
4          1  superbly trashy and wondrously unpretentious 8...

In [33]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [34]:
#Creating a function to remove stopwords [The,is,and] these  doesnot contribute to the analysis

In [35]:
def remove_stopwords(text):
    x=[]
    for i in text.split():
        if i not in stopwords.words('english'):
            x.append(i)
    y=x[:]
    x.clear()
    return y

In [36]:
data['review'] = data['review'].apply(remove_stopwords)

In [38]:
data.head()

sentiment                                             review
0          1  [stuff, going, moment, mj, ive, started, liste...
1          1  [classic, war, worlds, timothy, hines, enterta...
2          0  [film, starts, manager, nicholas, bell, giving...
3          0  [must, assumed, praised, film, greatest, filme...
4          1  [superbly, trashy, wondrously, unpretentious, ...

In [39]:
# Now we will have to go through stemming process

In [40]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

In [41]:
#Function for removing stem words

In [42]:
y=[]
def stem_words(text):
    for i in text:
        y.append(ps.stem(i))
    z=y[:]
    y.clear()
    return z

In [43]:
data['review'] = data['review'].apply(stem_words)

In [44]:
data.head()

sentiment                                             review
0          1  [stuff, go, moment, mj, ive, start, listen, mu...
1          1  [classic, war, world, timothi, hine, entertain...
2          0  [film, start, manag, nichola, bell, give, welc...
3          0  [must, assum, prais, film, greatest, film, ope...
4          1  [superbl, trashi, wondrous, unpretenti, 80, ex...

In [45]:
def join_back(list_input):
    return " ".join(list_input)

In [46]:
data['review'] = data['review'].apply(join_back)

FEATURE EXTRACTION

In [47]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer # Import CountVectorizer

In [48]:
vectorizer = CountVectorizer()

In [53]:
X = vectorizer.fit_transform(data.review)

In [54]:
y = data.sentiment

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [57]:
X_train.shape

(3073, 23525)

In [58]:
X_train.shape

(3073, 23525)

In [59]:
y_train.shape

(3073,)

In [60]:
y_train

2934    1
2423    1
433     1
2042    1
371     1
       ..
1130    0
1294    1
860     1
3507    0
3174    1
Name: sentiment, Length: 3073, dtype: int64

# Starting with Traditional Machine Learning models

Logistic Regression

In [61]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [62]:
model = LogisticRegression()

In [63]:
model.fit(X_train, y_train)

LogisticRegression()

In [64]:
y_pred = model.predict(X_test)

In [65]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8023407022106632


Using Naive Bayes

In [66]:
from sklearn.naive_bayes import GaussianNB, MultinomialNB

In [67]:
clf1 = GaussianNB()
clf2 = MultinomialNB()

In [68]:
clf1.fit(X_train.toarray(), y_train)
clf2.fit(X_train, y_train)

MultinomialNB()

In [69]:
accuracy1 = clf1.score(X_test.toarray(), y_test)
accuracy2 = clf2.score(X_test, y_test)

In [70]:
print("Accuracy of Gaussian Naive Bayes:", accuracy1)
print("Accuracy of Multinomial Naive Bayes:", accuracy2)

Accuracy of Gaussian Naive Bayes: 0.6280884265279584
Accuracy of Multinomial Naive Bayes: 0.8153446033810143


Now using SVM Model

In [71]:
from sklearn import svm

In [72]:
clf = svm.SVC(kernel='linear')

In [73]:
clf.fit(X_train, y_train)

SVC(kernel='linear')

In [74]:
y_pred = clf.predict(X_test)

In [75]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.7750325097529259


Now using Random Forest

In [76]:
from sklearn.ensemble import RandomForestClassifier

In [77]:
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

In [78]:
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [79]:
y_pred = rf_model.predict(X_test)

In [80]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8023407022106632


# DEEP Learning Methods

Using RNN

In [92]:
!pip install tensorflow

In [93]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense

In [94]:
X_train_reshaped = X_train.toarray().reshape(X_train.shape[0], 1, X_train.shape[1])
X_test_reshaped = X_test.toarray().reshape(X_test.shape[0], 1, X_test.shape[1])

In [96]:
model = Sequential()
model.add(SimpleRNN(units=32, input_shape=(1, X_train.shape[1])))
model.add(Dense(units=1, activation='sigmoid'))

In [97]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [98]:
model.fit(X_train_reshaped, y_train, epochs=10, batch_size=32, validation_data=(X_test_reshaped, y_test))

Epoch 1/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 37s 25ms/step - accuracy: 0.6992 - loss: 0.5877 - val_accuracy: 0.8283 - val_loss: 0.4111
Epoch 2/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9504 - loss: 0.1995 - val_accuracy: 0.8336 - val_loss: 0.3966
Epoch 3/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9872 - loss: 0.0883 - val_accuracy: 0.8179 - val_loss: 0.4223
Epoch 4/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.9944 - loss: 0.0441 - val_accuracy: 0.8192 - val_loss: 0.4484
Epoch 5/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.9984 - loss: 0.0242 - val_accuracy: 0.8270 - val_loss: 0.4792
Epoch 6/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.9994 - loss: 0.0154 - val_accuracy: 0.8218 - val_loss: 0.5003
Epoch 7/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9992 - loss: 0.0118 - val_accuracy: 0.8205 - val_loss: 0.5230
Epoch 8/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.9998 - loss: 0.0086 - val_accuracy: 0.8153 - val_los

In [99]:
loss,accuracy = model.evaluate(X_test_reshaped, y_test)

25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8083 - loss: 0.5968


In [100]:
print("Test Loss:", loss)
print("Test Accuracy:", accuracy)

Test Loss: 0.5748693346977234
Test Accuracy: 0.8101430535316467


Using LSTM

In [109]:
!pip install tensorflow

In [113]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
import scipy.sparse as sp

In [ ]:
#Reshape the data for LSTM

In [115]:
# Converting sparse matrix to dense NumPy array
X_train_dense = X_train.toarray()

# Reshaping the data for LSTM
x_train_reshaped = X_train_dense.reshape(X_train_dense.shape[0], 1, X_train_dense.shape[1])

In [119]:
model = Sequential()
model.add(LSTM(units=64, input_shape=(1, X_train.shape[1])))  # Adjusted units as needed
model.add(Dense(1, activation='sigmoid'))

In [120]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [121]:
model.fit(x_train_reshaped, y_train, epochs=10, batch_size=32)  # Adjusted epochs and batch size

Epoch 1/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.7050 - loss: 0.6173
Epoch 2/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - accuracy: 0.9457 - loss: 0.2012
Epoch 3/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step - accuracy: 0.9855 - loss: 0.0710
Epoch 4/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step - accuracy: 0.9963 - loss: 0.0317
Epoch 5/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 9ms/step - accuracy: 0.9985 - loss: 0.0152
Epoch 6/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.9985 - loss: 0.0105
Epoch 7/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 2s 13ms/step - accuracy: 0.9998 - loss: 0.0075
Epoch 8/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.9999 - loss: 0.0055
Epoch 9/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 0.9994 - loss: 0.0044
Epoch 10/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - accuracy: 1.0000 - loss: 0.0030


In [131]:
from scipy.sparse import csr_matrix
import numpy as np

# Reshaping test data to 2D for compatibility with csr_matrix
X_test_dense = X_test.toarray()
X_test_reshaped = csr_matrix(X_test_dense.reshape(X_test_dense.shape[0], -1))

# Converting sparse matrix to dense array
X_test_dense_final = X_test_reshaped.toarray()

# Reshaping to match the expected input shape of the model
X_test_reshaped_final = X_test_dense_final.reshape(X_test_dense_final.shape[0], 1, -1)

loss, accuracy = model.evaluate(X_test_reshaped_final, y_test)
print("Loss:", loss)
print("Accuracy:", accuracy)

25/25 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8036 - loss: 0.7607
Loss: 0.7054948806762695
Accuracy: 0.8140442371368408


Using CNN

In [132]:
# Reshaping Data for CNN
vocab_size = len(vectorizer.vocabulary_)
max_length = X_train.shape[1]  # Maximum length of a review (number of words)

X_train = X_train.toarray().reshape(X_train.shape[0], max_length, 1)
X_test = X_test.toarray().reshape(X_test.shape[0], max_length, 1)


In [145]:
import keras

In [151]:
from keras.models import Sequential
from keras.layers import Embedding, Conv1D, GlobalMaxPooling1D, Dense , Reshape

In [152]:
# Building the CNN Model
model = Sequential()
model.add(Embedding(vocab_size, 128, input_length=max_length))  # Embedding layer

# Reshape the output of the Embedding layer to be compatible with Conv1D
model.add(Reshape((max_length, 128)))

model.add(Conv1D(128, 5, activation='relu'))  # Convolutional layer
model.add(GlobalMaxPooling1D())  # Pooling layer
model.add(Dense(1, activation='sigmoid'))  # Output layer

# Compiling the Model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [153]:
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))

Epoch 1/10


97/97 ━━━━━━━━━━━━━━━━━━━━ 32s 249ms/step - accuracy: 0.5036 - loss: 0.6958 - val_accuracy: 0.5332 - val_loss: 0.6910
Epoch 2/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 29s 189ms/step - accuracy: 0.5016 - loss: 0.6973 - val_accuracy: 0.4655 - val_loss: 0.7002
Epoch 3/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 19s 194ms/step - accuracy: 0.5012 - loss: 0.6921 - val_accuracy: 0.5254 - val_loss: 0.6930
Epoch 4/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 19s 193ms/step - accuracy: 0.5199 - loss: 0.6942 - val_accuracy: 0.4941 - val_loss: 0.6944
Epoch 5/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 18s 189ms/step - accuracy: 0.5219 - loss: 0.6914 - val_accuracy: 0.4954 - val_loss: 0.6937
Epoch 6/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 21s 194ms/step - accuracy: 0.5075 - loss: 0.6935 - val_accuracy: 0.4967 - val_loss: 0.6943
Epoch 7/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 20s 193ms/step - accuracy: 0.5270 - loss: 0.6899 - val_accuracy: 0.4928 - val_loss: 0.6940
Epoch 8/10
97/97 ━━━━━━━━━━━━━━━━━━━━ 20s 193ms/step - accuracy: 0.5306 - loss: 0.6911 - val_accuracy: 0.487

In [156]:
accuracy = model.evaluate(X_test, y_test)[0]
print("Accuracy:", accuracy)

25/25 ━━━━━━━━━━━━━━━━━━━━ 1s 38ms/step - accuracy: 0.5288 - loss: 0.6893
Accuracy: 0.6893922686576843


**Traditional Machine Learning Methods Used are**

1.   Logistic regression with 0.88 Accuracy [First Read(Actual) ]
2.   Naive Bayes with 0.81 Accuracy
3.   Support Vector Machine with 0.77 Accuracy
4.   Random Forest [Ensemble] with 0.80 Accuracy


**Deep Learning Techniques Used Are **
1. RNN with 0.81 Accuracy
2. LSTM with 0.82 Accuracy
3. CNN with 0.68 Accuracy